src/signature.rs

```diff
@@ -315,22 +315,50 @@ fn build_canonical_query_string(params: &Params) -> String {
         if !output.is_empty() {
             output.push_str("&");
         }
-        output.push_str(&byte_serialize(item.0));
+        output.push_str(&encode_uri_strict(item.0));
         output.push_str("=");
-        output.push_str(&byte_serialize(item.1));
+        output.push_str(&encode_uri_strict(item.1));
     }
 
     output
 }
 
+#[derive(Clone)]
+pub struct StrictEncodeSet;
+
+impl EncodeSet for StrictEncodeSet {
```

Comment:
Let's add a comment in here explaining why we're doing this.  A copy/paste of the relevant section of http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html would be fine:

```rust
// Do not URI-encode any of the unreserved characters that RFC 3986 defines: A-Z, a-z, 0-9, hyphen ( - ), underscore ( _ ), period ( . ), and tilde ( ~ ).
// Percent-encode all other characters with %XY, where X and Y are hexadecimal characters (0-9 and uppercase A-F). For example, the space character must be encoded as %20 (not using '+', as some encoding schemes do) and extended UTF-8 characters must be in the form %XY%ZA%BC.
```